mat/symband.go

```diff
@@ -0,0 +1,127 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+var (
+	symBandDense *SymBandDense
+	_            Matrix           = symBandDense
+	_            Symmetric        = symBandDense
+	_            Banded           = symBandDense
+	_            RawSymBander     = symBandDense
+	_            MutableSymBanded = symBandDense
+)
+
+// SymBandDense represents a symmetric band matrix in dense storage format.
+type SymBandDense struct {
+	mat blas64.SymmetricBand
+}
+
+// MutableSymBanded is a symmetric band matrix interface type that allows elements
+// to be altered.
+type MutableSymBanded interface {
+	Symmetric
+	Bandwidth() (kl, ku int)
+	SetSymBand(i, j int, v float64)
+}
+
+// A RawSymBander can return a blas64.SymmetricBand representation of the receiver.
+// Changes to the blas64.SymmetricBand.Data slice will be reflected in the original
+// matrix, changes to the N, K, Stride and Uplo fields will not.
+type RawSymBander interface {
+	RawSymBand() blas64.SymmetricBand
+}
+
+// NewSymBandDense creates a new SymBand matrix with n rows and columns. If data == nil,
+// a new slice is allocated for the backing slice. If len(data) == n*(k+1),
+// data is used as the backing slice, and changes to the elements of the returned
+// BandDense will be reflected in data. If neither of these is true, NewSymBandDense
+// will panic. k must be at least zero and less than n, otherwise NewBandDense will panic.
+//
+// The data must be arranged in row-major order constructed by removing the zeros
+// from the rows outside the band and aligning the diagonals. SymBandDense matrices
+// are stored in the upper triangle. For example, the matrix
+//    1  2  3  0  0  0
+//    2  4  5  6  0  0
+//    3  5  7  8  9  0
+//    0  6  8 10 11 12
+//    0  0  9 11 13 14
+//    0  0  0 12 14 15
+// becomes (* entries are never accessed)
```

Comment:
I see your point, but ... 

We have to do this because this is the C interface definition? For the other types, the place of the elements in the matrix is the same, they're just in a different order in the linear array because of row major. I think it's going to be much easier to have our layout match Lapack, rather than having to understand the effects of these new quasi-transposes for every new function. I would happily re-write our BLAS implementation to make this the case (so bands are still down the row), but I don't know how this would interact with the C wrappers.